fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/HudiScanPartitionPruneRule.java

```diff
@@ -0,0 +1,224 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+public class HudiScanPartitionPruneRule extends TransformationRule {
+    private static final Logger LOG = LogManager.getLogger(HudiScanPartitionPruneRule.class);
+
+    public HudiScanPartitionPruneRule() {
+        super(RuleType.TF_PARTITION_PRUNE, Pattern.create(OperatorType.LOGICAL_HUDI_SCAN));
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        LogicalHudiScanOperator operator = (LogicalHudiScanOperator) input.getOp();
+        if (operator.getTableType() == Table.TableType.HUDI) {
+            // partitionColumnName -> (LiteralExpr -> partition ids)
+            // no null partitions in this map, used by ListPartitionPruner
+            Map<ColumnRefOperator, TreeMap<LiteralExpr, Set<Long>>> columnToPartitionValuesMap = Maps.newHashMap();
+            // Store partitions with null partition values separately, used by ListPartitionPruner
+            // partitionColumnName -> null partitionIds
+            Map<ColumnRefOperator, Set<Long>> columnToNullPartitions = Maps.newHashMap();
+
+            try {
+                initPartitionInfo(operator, columnToPartitionValuesMap, columnToNullPartitions, context);
+                classifyConjuncts(operator, columnToPartitionValuesMap);
+                computePartitionInfo(operator, columnToPartitionValuesMap, columnToNullPartitions, context);
+            } catch (Exception e) {
+                LOG.warn("Hive table partition prune failed : " + e);
+                throw new StarRocksPlannerException(e.getMessage(), ErrorType.INTERNAL_ERROR);
+            }
+        }
+        return Collections.emptyList();
+    }
+
+    private void initPartitionInfo(LogicalHudiScanOperator operator,
+                                   Map<ColumnRefOperator, TreeMap<LiteralExpr, Set<Long>>> columnToPartitionValuesMap,
+                                   Map<ColumnRefOperator, Set<Long>> columnToNullPartitions,
+                                   OptimizerContext context) throws DdlException {
+        HudiTable hiveTable = (HudiTable) operator.getTable();
+        List<Column> partitionColumns = hiveTable.getPartitionColumns();
+        List<ColumnRefOperator> partitionColumnRefOperators = new ArrayList<>();
+        for (Column column : partitionColumns) {
+            ColumnRefOperator partitionColumnRefOperator = operator.getColumnReference(column);
+            columnToPartitionValuesMap.put(partitionColumnRefOperator, new TreeMap<>());
+            columnToNullPartitions.put(partitionColumnRefOperator, Sets.newHashSet());
+            partitionColumnRefOperators.add(partitionColumnRefOperator);
+        }
+
+        // no partition column table:
+        // 1. partitionColumns is empty
+        // 2. partitionKeys size = 1
+        // 3. key.getKeys() is empty
+        Map<PartitionKey, Long> partitionKeys = hiveTable.getPartitionKeys();
```

Comment:
nit: hiveTable -> hudiTable?
and i think we should add a common rule for hive and hudi since many codes are repeated

Reply:
done, As the PR description said, I will consider these refactor in the following PRs because there are many places that need to be modified, which will interfere review process of the PR.